### PR TITLE
Manual QA Guidelines

### PR DESCRIPTION
**Pull request by mmislam101**

### CODE REVIEW

**Comment by mmislam101** [comment]
@swilliams I can't add Rae as a reviewer or tag her on these PRs
@krantzinator

**Comment by jesdaigle** [comment]
I don’t think she has accepted the invitations yet. Github says there are 3 pending. 


Thank you,
Jes Daigle
(802) 582-9207
 
———————————————————————————————————————————
LinkedIn: www.linkedin.com/in/jesdaigle <http://www.linkedin.com/in/jesdaigle>
Blog: https://medium.com/@techandautism <https://medium.com/@techandautism>
Tech Ladies Founding Member 
FindMyFlock mentor
Writer for HackerNoon

Life in permanent beta.

**Comment by mmislam101** [comment]
She's able to go in and review PRs though...

**Comment by jesdaigle** [comment]
Ask her to try again please. :)
Thank you,
Jes Daigle
(802) 582-9207
 
———————————————————————————————————————————
LinkedIn: www.linkedin.com/in/jesdaigle <http://www.linkedin.com/in/jesdaigle>
Blog: https://medium.com/@techandautism <https://medium.com/@techandautism>
Tech Ladies Founding Member 
FindMyFlock mentor
Writer for HackerNoon

Life in permanent beta.

File: docs/Quality.md [hunk 1]
```diff
@@ -103,3 +112,40 @@ This process can happen automatically with [GitLab's CI pipelines](https://docs.
...
+- Has it been tested before releasing?
+- What happens if a previous release needs to be patched?
+- What versions of software components integrate with each other?
+
+### Software Integration
+Software rarely comes in singles. There is usually at least a client side application that pairs with server side code. If done well, the server code should always remain backwards compatible. Meaning, any updates to the server should continue to maintain previous versions of client application that users haven't updated yet.
+
+It's still a good idea to capture and release software in groups. Especially when you have multiple types of client software, such as a dashboard that pairs with a client application, and they both share the same server backend.
+
+The best approach is to add **git tags** with [Semantic Versioning](https://semver.org/) to each of the software components being integrated. The tags will make it easy for developers to roll back to the exact software for that version.
```
Comment (krantzinator) [inline docs/Quality.md]: link to something like https://git-scm.com/book/en/v2/Git-Basics-Tagging to explain what `git tags` are

Comment (krantzinator) [reply docs/Quality.md]: what do you think about creating releases for each major/Named version?
https://help.github.com/articles/creating-releases/

**Comment by mmislam101** [comment]
@krantzinator updated. Good suggestions

File: docs/Quality.md [hunk 2]
```diff
@@ -94,13 +94,13 @@ Good details in a bug report include:
   * Trivial (eg. a copy error)
 
 ### Agile Manual QA Overflow
-Developing automated unit or UI tests are time consuming. It's almost impossible to maintain automation tests while a feature is being developed on at the same time. And often having enough time to develop unit tests and UI automation tests while developing is a luxury we don't get. So, to continue to maintain high quality, manual QA becomes necessary until the feature settles down.
+Developing automated unit or UI tests are time consuming. It's almost impossible to maintain automation tests while a feature is being developed. And often, having enough time to create tests while developing is a luxury we don't get. So, to continue to maintain high quality, manual QA becomes necessary until the feature settles down.
```
Comment (krantzinator) [inline docs/Quality.md]: this looks good, although i would like to discuss our testing goals in more detail as a team, specifically in regards to "creat[ing] tests while developing is a luxury we don't get."

File: docs/Quality.md [hunk 3]
```diff
@@ -94,13 +94,13 @@ Good details in a bug report include:
...
 
-Once there is time, and the churn on a feature has stopped, the items from the manual QA document should be converted to either unit tests, or UI automation tests. This creates an *overflow* type of process, like the coolant overflow tank in a car. As you develop and things are hot, new tests should "overflow" to the manual QA document, and then once development has cooled off, items should be removed from the manual QA document and converted to Unit or UI Automation tests.
+Once there is time, and churn on a feature ends, the items from the manual QA document should be converted to either unit tests, or UI automation tests. This creates an *overflow* process, like the coolant overflow tank in a car. As you develop, and things are hot, new tests should "overflow" to the manual QA document. Then, once development has cooled off, items should be removed from the manual QA document and converted to Unit or UI Automation tests.
```
Comment (krantzinator) [inline docs/Quality.md]: this sounds intriguing. i think following a BDD pattern for these manual tests could make it easier to convert them into acceptance tests. i'll submit a PR for short explanations of a few main testing strategies, which includes BDD.

File: docs/Quality.md (hunk 2 above)
Comment (mmislam101) [reply docs/Quality.md]: Think I'm wrong to mention unit tests as a luxury. I'll remove unit tests out of that luxury and replace it with integration tests.

File: docs/Quality.md [hunk 4]
```diff
@@ -94,12 +94,58 @@ Good details in a bug report include:
   * Trivial (eg. a copy error)
 
 ### Bug Triage
-Bugs should be treated similarly as issues and scheduled as a part of Sprint grooming and planning. Showstoppers and certain Major issues will likely require immediate attention. All other severities should be mixed in with stories and scheduled appropriately.
+Bugs should be treated similarly to issues and scheduled as a part of Sprint grooming and planning. Showstoppers and certain Major issues will likely require immediate attention. All other severities should be mixed in with stories and scheduled appropriately.
 
-## Continuous Integration
+### Agile Manual QA Overflow
+Developing automated integration or UI tests are time consuming. It's almost impossible to maintain automation tests while a feature is being developed. And often, having enough time to create UI and integration tests while developing is a luxury we don't get. Even though your development process might include TDD, unit tests won't catch the bugs that occur due to integration. So, to continue to maintain high quality, manual QA becomes necessary until the feature settles down.
```
Comment (swilliams) [inline docs/Quality.md]: "It's almost impossible to maintain automation tests while a feature is being developed." I disagree with this sentence.

File: docs/Quality.md [hunk 5]
```diff
@@ -94,12 +94,58 @@ Good details in a bug report include:
...
+
+To minimize human error, but maximize time to release, the best approach is to develop a Manual QA Document, or checklist, that captures items that should be tested before a release. These items should include both the normal vectors, and also corner case vectors of the application. This document should be updated regularly as more corner cases appear, bugs are found and squashed, and features are changed.
+
+Once there is time, and churn on a feature ends, the items from the manual QA document should be converted to either integration or UI automation tests. This creates an *overflow* process, like the coolant overflow tank in a car. As you develop, and things are hot, new tests should "overflow" to the manual QA document. Then, once development has cooled off, items should be removed from the manual QA document and converted to Unit or UI Automation tests.
+
+The manual QA checklist is only useful if there's a process to go through manual QA before releasing the product.
```
Comment (swilliams) [inline docs/Quality.md]: This is generally referred to as a "script" in QA parlance I believe.

File: docs/Quality.md [hunk 6]
```diff
@@ -94,12 +94,58 @@ Good details in a bug report include:
...
+
+## Release
+A good release process is essential in maintaining quality. It's important to consider:
+- What is going into a release?
+- Has it been tested before releasing?
+- What happens if a previous release needs to be patched?
+- What versions of software components integrate with each other?
```
Comment (swilliams) [inline docs/Quality.md]: + `- How to back out if things go sideways?`

File: docs/Quality.md [hunk 7]
```diff
@@ -94,12 +94,58 @@ Good details in a bug report include:
...
+| Orangutan     | 1.1.6     | 1.2.8   | 1.1.5  |
+
+### Patching Releases
+Development never stops, and software is never perfect. Often, by the time a bug is found on a released software, develpment has already added hundreds of commits on the master branch. This is when the power of **git tags** shine.
+1. Create a patch release branch off the tag of the release that needs to be fixed. Name it something like: `patch-1.1.6` (where `1.1.6` is a new version from the released version of `1.1.5`)
+2. PR the fix onto this patch release branch.
+3. Once merged, create a new PR to implement the same fix on the master branch. Or cherry-pick it directly onto the master branch. This will ensure that the next major release will have this fix already integrated.
+4. Keep the patch release branch around. You can create a new tag, but this tag won't be helpful. Tags are associated with a commit. So when rolling back to a tag, it necessarily won't have the same git chain as the patch release branch, especially when cherry-picking the commit from the patch branch to on top of the master branch.
+5. Create a new entry in the software integration document:
```
Comment (swilliams) [inline docs/Quality.md]: Where does the document live?

File: docs/Quality.md (hunk 4 above)
Comment (mmislam101) [reply docs/Quality.md]: How about "hard to maintain"?

Comment (swilliams) [reply docs/Quality.md]: That works.

**Pull request merged**